adr/ADR-17.md

```diff
@@ -0,0 +1,48 @@
+# Ordered Consumer
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-09-29|
+|Author  |@scottf|
+|Status  |Proposed|
+|Tags    |jetstream,client|
+
+### Context and Problem Statement
+
+Provide an ordered push subscription for the user that automatically checks and recovers when a gap occurs in the consumer sequence.
+The subscription can deliver messages synchronously or asynchronously as normally supported by the client.
+
+### Behavior
+
+The subscription should leverage Gap Management and Auto Status Management to ensure messages are received in the proper order.
+
+The subscription must track the last good stream and consumer sequences.
+When a gap is observed, the subscription closes its current subscription,
+releases its consumer and creates a new one starting at the proper stream sequence.
+
+You can optionally make the "state" available to the user.
+
+### Subscription Limitations
+
+The subscription cannot be for 
+- a pull consumer
+- a durable consumer
+- cannot bind or be "direct"
+
+The subscription is not allowed with queues/deliver groups.
+
+### Consumer Configuration Checks
+
+The user can provide a consumer configuration but it must be validated. Error when creating if validation fails.
+
+Checks:
+
+- durable_name: can not be provided
```

Comment:
You removed the deliver subject check, but it was correct to say that it is not allowed. In go, we have this check:
```
// No deliver subject, we pick our own.
if o.cfg.DeliverSubject != _EMPTY_ {
    return nil, fmt.Errorf("nats: deliver subject can not be set for an ordered consumer")
}
```
The reason is that we want the ephemeral created to have a unique (inbox) deliver subject, which helps when doing a "reset" where the old one is stopped and new one created. Having the same subject may be an issue, I think.

Reply:
PS: I missed that one in the current C PR, so will add to it.

Reply:
ok I can add that instruction back in.